### PR TITLE
O365 connector validation & code updates

### PR DESCRIPTION
**Pull request by LucGenetier**
- Full validation of O365 Groups connector
- Full validation of O365 Users connector
- Fix in connector code around internal parameters (discovered a variation w/ PA Client which is confirmed as a bug on their end)
- add logic in HttpFunctionInvoker to trace requests (disabled by default)
- add support in test code for providing JPG response files as byte array

### CODE REVIEW

File: src/libraries/Microsoft.PowerFx.Core/Functions/ISupportsArgumentSuggestions.cs [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+using System.Collections.Generic;
+using Microsoft.PowerFx.Core.Types;
+using static Microsoft.PowerFx.Intellisense.ArgumentSuggestions;
+
+namespace Microsoft.PowerFx.Core.Functions
+{
+    // Used for Intellisense in connectors
+    internal interface ISupportsArgumentSuggestions
+    {
+        IEnumerable<KeyValuePair<string, DType>> GetArgumentSuggestions(TryGetEnumSymbol tryGetEnumSymbol, bool suggestUnqualifiedEnums, DType scopeType, int argumentIndex, out bool requiresSuggestionEscaping);
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Core/Functions/ISupportsArgumentSuggestions.cs:13]: >string [](http://example.com/codeflow?start=33&length=6)

can we document what this is? 

What exactly is string and Dtype here?

File: src/libraries/Microsoft.PowerFx.Connectors/Execution/HttpFunctionInvoker.cs [hunk 2]
```diff
@@ -303,13 +303,48 @@ private HttpContent GetBody(string referenceId, bool schemaLessBody, Dictionary<
...
         public async Task<FormulaValue> DecodeResponseAsync(HttpResponseMessage response, bool throwOnError = false)
-        {
+        {            
             var text = response?.Content == null
                             ? string.Empty
                             : await response.Content.ReadAsStringAsync().ConfigureAwait(false);
 
             var statusCode = (int)response.StatusCode;
 
+#if RECORD_RESULTS
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Connectors/Execution/HttpFunctionInvoker.cs:313]: Is this intended to check in; or just temporary debugging code?

File: src/libraries/Microsoft.PowerFx.Core/Texl/Intellisense/IntellisenseHelper.cs [hunk 3]
```diff
@@ -304,15 +304,16 @@ internal static void AddSuggestionsForNamespace(IntellisenseData.IntellisenseDat
         /// <summary>
         /// Adds suggestions that start with the MatchingString from the given type.
         /// </summary>
-        internal static void AddTopLevelSuggestions(IntellisenseData.IntellisenseData intellisenseData, DType type, string prefix = "")
+        internal static void AddTopLevelSuggestions(IntellisenseData.IntellisenseData intellisenseData, DType type, string prefix = "", string filter = null)
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Core/Texl/Intellisense/IntellisenseHelper.cs:307]: > string filter = null [](http://example.com/codeflow?start=135&length=21)

what is this for?

File: src/tests/Microsoft.PowerFx.Connectors.Tests/Responses/Response_O365Users_TrendingDocuments.json [hunk 4]
```diff
@@ -0,0 +1 @@
...t":"Erererer\r\ngghfgghghghhg\r\n","title":"testdoc1","type":"Word"}},{"id":"TF_sfTnPPlgjEG8h19PyvRwWMQEkN2EO9NLt7OOXQRVFafNtEQWr9hYRbuRGKZLJ85MwtKC7c9mHUiJfaCNANizdA","weight":0.0,"resourceReference":{"id":"drives/b!sfTnPPlgjEG8h19PyvRwWMQEkN2EO9NLt7OOXQRVFafNtEQWr9hYRbuRGKZLJ85M/items/01ITWY52WC2KBO3T3GDVEIS7NARUANRM3U","type":"microsoft.graph.driveItem","webUrl":"https://capintegration01.sharepoint.com/_layouts/15/Doc.aspx?sourcedoc=%7BED82D2C2-66CF-481D-897D-A08D00D8B374%7D&file=testdoc1.docx&action=default&mobileredirect=true&DefaultItemOpen=1"},"resourceVisualization":{"containerDisplayName":"Communication site","containerType":"Site","containerWebUrl":"https://capintegration01.sharepoint.com/account/testacct1_0D49CAE1AE84EE118178000D3A9E1D40","mediaType":"application/vnd.openxmlformats-officedocument.wordprocessingml.document","previewImageUrl":"https://capintegration01.sharepoint.com/_api/v2.1/drives/b!sfTnPPlgjEG8h19PyvRwWMQEkN2EO9NLt7OOXQRVFafNtEQWr9hYRbuRGKZLJ85M/items/01ITWY52WC2KBO3T3GDVEIS7NARUANRM3U/thumbnails/0/small/content?prefer=noredirect,closestavailablesize","previewText":"Dsjkdsjksdjk\r\nSdkjsdjksdjksd\r\nsdkljksdkjsd\r\n","title":"testdoc1","type":"Word"}}]}
```
Comment (MikeStall) [inline src/tests/Microsoft.PowerFx.Connectors.Tests/Responses/Response_O365Users_TrendingDocuments.json:1]: super nit - can you format this? 
single files with very long lines don't play well with git/tooling

**Comment by MikeStall** [comment]
Do we need to check in the full jpg? 

src/tests/Microsoft.PowerFx.Connectors.Tests/Responses/Response_O365Users_UserPhoto.jpeg

Are we actually using and verifying this?

**Comment by LucGenetier** [comment]
Public API changes: <br />❌CP0002 M:Microsoft.PowerFx.Connectors.OpenApiExtensions.GetReturnType(Microsoft.OpenApi.Models.OpenApiOperation)

File: src/libraries/Microsoft.PowerFx.Core/Texl/Intellisense/IntellisenseHelper.cs [hunk 5]
```diff
@@ -304,15 +304,16 @@ internal static void AddSuggestionsForNamespace(IntellisenseData.IntellisenseDat
...
         {
             Contracts.AssertValue(intellisenseData);
             Contracts.Assert(type.IsValid);
             Contracts.Assert(prefix.Length == 0 || (TexlLexer.PunctuatorBang + TexlLexer.PunctuatorDot).IndexOf(prefix[prefix.Length - 1]) >= 0);
 
             foreach (var tName in type.GetRootFieldNames().Select(name => (Type: type.GetType(name), Name: name)))
             {
-                if (!intellisenseData.TryAddCustomColumnTypeSuggestions(tName.Type))
+                bool filtered = string.IsNullOrEmpty(filter) ? false : !tName.Name.Value.StartsWith(filter, StringComparison.OrdinalIgnoreCase);
```
Comment (jas-valgotar) [inline src/libraries/Microsoft.PowerFx.Core/Texl/Intellisense/IntellisenseHelper.cs:315]: >bool filtered = string.IsNullOrEmpty(filter) ? false : !tName.Name.Value.StartsWith(filter, StringComparison.OrdinalIgnoreCase); [](http://example.com/codeflow?start=16&length=128)

Will this also prevent us from suggesting substring matches? is that intended? 
ie. If I type `cat`, should I not get `Concat` in suggestion?

File: src/libraries/Microsoft.PowerFx.Connectors/OpenApiExtensions.cs [hunk 6]
```diff
@@ -573,9 +578,9 @@ public static HttpMethod ToHttpMethod(this OperationType key)
             };
         }
 
-        public static FormulaType GetReturnType(this OpenApiOperation openApiOperation)
+        public static FormulaType GetReturnType(this OpenApiOperation openApiOperation, ConnectorCompatibility compatibility)
```
Comment (jas-valgotar) [inline src/libraries/Microsoft.PowerFx.Connectors/OpenApiExtensions.cs:581]: >public static FormulaType GetReturnType(this OpenApiOperation openApiOperation, ConnectorCompatibility compatibility) [](http://example.com/codeflow?start=8&length=117)

do we need to worry abt this breaking change? I'll check if Plugin uses this...

File: src/libraries/Microsoft.PowerFx.Core/Texl/Intellisense/IntellisenseHelper.cs [hunk 7]
```diff
@@ -355,14 +356,38 @@ public static void AddTopLevelSuggestionsForCursorType(IntellisenseData.Intellis
...
+                        int start = node.Token.Span.Min;
+                        int end = node.Token.Span.Lim;
+
+                        // if node is BinaryOpNode, use the Right part for determining the filter
+                        if (node is BinaryOpNode binOpNode)
                         {
+                            start = binOpNode.Right.Token.Span.Min;
+                            end = binOpNode.Right.Token.Span.Lim;
+                        }
+
```
Comment (jas-valgotar) [inline src/libraries/Microsoft.PowerFx.Core/Texl/Intellisense/IntellisenseHelper.cs:371]: If there is specific logic to binaryopnode can we possibly incorporate this in TryAddSuggestionForCurrentBinaryOp?

**Review by MikeStall — CHANGES_REQUESTED** [review]
:clock1:

File: src/libraries/Microsoft.PowerFx.Core/Texl/Intellisense/IntellisenseHelper.cs [hunk 8]
```diff
@@ -355,14 +356,38 @@ public static void AddTopLevelSuggestionsForCursorType(IntellisenseData.Intellis
 
                     if (!intellisenseData.TryAddSuggestionForCurrentBinaryOp(checkForOptionSetOnly: true))
                     {
-                        if (info.ScopeIdentifier != default)
+                        var node = callNode.Args.Children[argPosition];
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Core/Texl/Intellisense/IntellisenseHelper.cs:359]: >                      var node = callNode.Args.Children[argPosition]; [](http://example.com/codeflow?start=2&length=69)

does PAClient have code like this that we can copy instead? 
We should understand how PAClient is working

File: src/libraries/Microsoft.PowerFx.Core/Texl/Intellisense/IntellisenseHelper.cs [hunk 9]
```diff
@@ -668,7 +693,7 @@ public static bool AddTopLevelSuggestionsForGivenNode(IntellisenseData.Intellise
                 }
             }
 
-            if (callNode.Parent != null)
+            if (callNode.Parent != null && node is not ErrorNode)
```
Comment (jas-valgotar) [inline src/libraries/Microsoft.PowerFx.Core/Texl/Intellisense/IntellisenseHelper.cs:696]: >&& node is not ErrorNode [](http://example.com/codeflow?start=40&length=24)

what kind of expression needs this condition? can you please give an example?

File: src/libraries/Microsoft.PowerFx.Connectors/OpenApiExtensions.cs (hunk 6 above)
Comment (anderson-joyle) [reply src/libraries/Microsoft.PowerFx.Connectors/OpenApiExtensions.cs:581]: nor cards

Comment (LucGenetier) [reply src/libraries/Microsoft.PowerFx.Connectors/OpenApiExtensions.cs:581]: Thanks for checking guys

File: src/libraries/Microsoft.PowerFx.Connectors/Execution/HttpFunctionInvoker.cs (hunk 2 above)
Comment (LucGenetier) [reply src/libraries/Microsoft.PowerFx.Connectors/Execution/HttpFunctionInvoker.cs:313]: I think it's safe to check in as this code will never end up in a nuget. 
It's only there for intercepting network results after a connector call & facilitate debugging/the creation of tests.

File: src/libraries/Microsoft.PowerFx.Core/Functions/ISupportsArgumentSuggestions.cs (hunk 1 above)
Comment (LucGenetier) [reply src/libraries/Microsoft.PowerFx.Core/Functions/ISupportsArgumentSuggestions.cs:13]: Removed

File: src/libraries/Microsoft.PowerFx.Core/Texl/Intellisense/IntellisenseHelper.cs (hunk 5 above)
Comment (LucGenetier) [reply src/libraries/Microsoft.PowerFx.Core/Texl/Intellisense/IntellisenseHelper.cs:315]: Removed

File: src/libraries/Microsoft.PowerFx.Core/Texl/Intellisense/IntellisenseHelper.cs (hunk 8 above)
Comment (LucGenetier) [reply src/libraries/Microsoft.PowerFx.Core/Texl/Intellisense/IntellisenseHelper.cs:359]: Removed

File: src/libraries/Microsoft.PowerFx.Core/Texl/Intellisense/IntellisenseHelper.cs [hunk 10]
```diff
@@ -355,14 +356,38 @@ public static void AddTopLevelSuggestionsForCursorType(IntellisenseData.Intellis
...
                     if (!intellisenseData.TryAddSuggestionForCurrentBinaryOp(checkForOptionSetOnly: true))
                     {
-                        if (info.ScopeIdentifier != default)
+                        var node = callNode.Args.Children[argPosition];
+
+                        // filter on what's already typed (up to the cursor position)
+                        int start = node.Token.Span.Min;
+                        int end = node.Token.Span.Lim;
+
+                        // if node is BinaryOpNode, use the Right part for determining the filter
```
[the remark this answers is not in the document]
Comment (LucGenetier) [reply src/libraries/Microsoft.PowerFx.Core/Texl/Intellisense/IntellisenseHelper.cs:365]: Removed

File: src/libraries/Microsoft.PowerFx.Core/Texl/Intellisense/IntellisenseHelper.cs (hunk 7 above)
Comment (LucGenetier) [reply src/libraries/Microsoft.PowerFx.Core/Texl/Intellisense/IntellisenseHelper.cs:371]: Removed

File: src/libraries/Microsoft.PowerFx.Core/Texl/Intellisense/IntellisenseHelper.cs (hunk 9 above)
Comment (LucGenetier) [reply src/libraries/Microsoft.PowerFx.Core/Texl/Intellisense/IntellisenseHelper.cs]: Removed

**Comment by LucGenetier** [comment]
Public API changes: <br />❌CP0002 M:Microsoft.PowerFx.Connectors.OpenApiExtensions.GetReturnType(Microsoft.OpenApi.Models.OpenApiOperation)

**Comment by LucGenetier** [comment]
Public API changes: <br />❌CP0002 M:Microsoft.PowerFx.Connectors.OpenApiExtensions.GetReturnType(Microsoft.OpenApi.Models.OpenApiOperation)

File: src/libraries/Microsoft.PowerFx.Connectors/Texl/ConnectorTexlFunction.cs [hunk 11]
```diff
@@ -127,5 +122,27 @@ public override async Task<ConnectorSuggestions> GetConnectorSuggestionsAsync(Fo
                 throw;
             }
         }
+
+        public IEnumerable<KeyValuePair<string, DType>> GetArgumentSuggestions(ArgumentSuggestions.TryGetEnumSymbol tryGetEnumSymbol, bool suggestUnqualifiedEnums, DType scopeType, int argumentIndex, out bool requiresSuggestionEscaping)
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Connectors/Texl/ConnectorTexlFunction.cs]: >GetArgumentSuggestions [](http://example.com/codeflow?start=56&length=22)

is this still used?

File: src/tests/Microsoft.PowerFx.Connectors.Tests/Microsoft.PowerFx.Connectors.Tests.csproj [hunk 12]
```diff
@@ -29,6 +29,7 @@
     <EmbeddedResource Include="Swagger\*.json" />
     <EmbeddedResource Include="Responses\*.json" />
     <EmbeddedResource Include="Responses\*.txt" />
+    <EmbeddedResource Include="Responses\*.jpeg" />
```
Comment (MikeStall) [inline src/tests/Microsoft.PowerFx.Connectors.Tests/Microsoft.PowerFx.Connectors.Tests.csproj:32]: reminder: remove

**Comment by LucGenetier** [comment]
Public API changes: <br />❌CP0002 M:Microsoft.PowerFx.Connectors.OpenApiExtensions.GetReturnType(Microsoft.OpenApi.Models.OpenApiOperation)

**Review by MikeStall — APPROVED** [review]

**Pull request merged**